src/app/server/JointFabricDatastore.cpp

```diff
@@ -127,185 +128,565 @@ CHIP_ERROR JointFabricDatastore::RemoveNode(NodeId nodeId)
+                                                                             entry.endpointID == entryToErase.endpointID &&
+                                                                             entry.groupID == entryToErase.groupID;
+                                                                     }),
+                                                      mEndpointGroupIDEntries.end());
+                    }));
+                }
+                else if (it->statusEntry.state == Clusters::JointFabricDatastore::DatastoreStateEnum::kCommitFailed)
+                {
+                    CHIP_ERROR failureCode(it->statusEntry.failureCode);
+
+                    if (failureCode == CHIP_IM_GLOBAL_STATUS(ConstraintError) ||
+                        failureCode == CHIP_IM_GLOBAL_STATUS(ResourceExhausted))
+                    {
+                        // remove entry from the list
+                        it = mEndpointGroupIDEntries.erase(it);
+                        continue;
+                    }
+                    else
+                    {
+                        // Retry the failed commit by attempting to SyncNode again.
+                        // Make a copy to send to the node. On success mark the entry Committed.
+                        // If SyncNode returns an error, leave the entry as CommitFailed so a later
+                        // Refresh can retry again.
+                        {
+                            size_t idx       = static_cast<size_t>(std::distance(mEndpointGroupIDEntries.begin(), it));
+                            auto entryToSync = *it;
+                            ReturnErrorOnFailure(mDelegate->SyncNode(mRefreshingNodeId, entryToSync, [this, idx]() {
+                                if (idx < mEndpointGroupIDEntries.size())
+                                {
+                                    mEndpointGroupIDEntries[idx].statusEntry.state =
+                                        Clusters::JointFabricDatastore::DatastoreStateEnum::kCommitted;
+                                }
+                            }));
+                        }
+                    }
                 }
             }
+
+            ++it;
         }
 
-        ++it;
-    }
+        // Request Endpoint Binding List from the device and transition to kRefreshingBindings.
+        ReturnErrorOnFailure(mDelegate->FetchEndpointBindingList(
+            mRefreshingNodeId,
+            [this](CHIP_ERROR err,
+                   const std::vector<Clusters::JointFabricDatastore::Structs::DatastoreEndpointBindingEntryStruct::Type> &
+                       endpointBindings) {
+                if (err == CHIP_NO_ERROR)
+                {
+                    // Store fetched endpoint-binding associations for processing in the next state.
+                    // TODO
+                    // mEndpointBindingList = endpointBindings;
```

Comment:
```suggestion
                    // mRefreshingEndpointBindingList = endpointBindings;
```

Naming of all temporary variables with "mRefreshing" might be helpful.